src/app/models/function.rb

```diff
@@ -6,11 +6,30 @@ class Function < ApplicationRecord
 
   belongs_to :user
 
+  before_validation :function_name
+
   def to_param
     name
   end
 
   def self.from_param(param)
     find_by(name: param)
   end
+
+  private
+
+  def walk_ast(ast)
+    return ast.children.first.to_s if ast.respond_to?('type') && ast.type == :DEFN
+
+    return unless ast.respond_to?('children')
+
+    ast.children.each do |child|
+      func_name = walk_ast(child)
+      return func_name unless func_name.nil?
```

Comment:
[FS] I think we can inverse this 
```suggestion
      return func_name if func_name
```